CodeGenService Crash on Back

1. Open turtle demo. Drag a block to the workspace.
1. Hit the play button in the action bar to invoke code generation.
1. Click the system **HOME** button (CIRCLE).
1. Click the system task button (SQUARE).
1. Find the Turtle demo. See the with screenshot including the workspace block. Select it.
1. Find the block still on the workspace.
1. Do not run code gen again.
1. Click the system **BACK** button (TRIANGLE). Click twice if the toolbox was open.
1. Crash (from API 24 develop branch):

```
E/AndroidRuntime: FATAL EXCEPTION: main
                  Process: com.google.blockly.demo, PID: 8742
                  java.lang.RuntimeException: Unable to pause activity {com.google.blockly.demo/com.google.blockly.android.demo.TurtleActivity}: java.lang.IllegalArgumentException: Service not registered: com.google.blockly.android.codegen.CodeGeneratorManager$1@86bad5c
                      at android.app.ActivityThread.performPauseActivityIfNeeded(ActivityThread.java:3722)
                      at android.app.ActivityThread.performPauseActivity(ActivityThread.java:3688)
                      at android.app.ActivityThread.performPauseActivity(ActivityThread.java:3662)
                      at android.app.ActivityThread.handlePauseActivity(ActivityThread.java:3636)
                      at android.app.ActivityThread.-wrap16(ActivityThread.java)
                      at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1481)
                      at android.os.Handler.dispatchMessage(Handler.java:102)
                      at android.os.Looper.loop(Looper.java:154)
                      at android.app.ActivityThread.main(ActivityThread.java:6077)
                      at java.lang.reflect.Method.invoke(Native Method)
                      at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:865)
                      at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:755)
                   Caused by: java.lang.IllegalArgumentException: Service not registered: com.google.blockly.android.codegen.CodeGeneratorManager$1@86bad5c
                      at android.app.LoadedApk.forgetServiceDispatcher(LoadedApk.java:1289)
                      at android.app.ContextImpl.unbindService(ContextImpl.java:1511)
                      at android.content.ContextWrapper.unbindService(ContextWrapper.java:648)
                      at com.google.blockly.android.codegen.CodeGeneratorManager.onPause(CodeGeneratorManager.java:80)
                      at com.google.blockly.android.BlocklyActivityHelper.onPause(BlocklyActivityHelper.java:214)
                      at com.google.blockly.android.AbstractBlocklyActivity.onPause(AbstractBlocklyActivity.java:285)
                      at android.app.Activity.performPause(Activity.java:6802)
                      at android.app.Instrumentation.callActivityOnPause(Instrumentation.java:1322)
                      at android.app.ActivityThread.performPauseActivityIfNeeded(ActivityThread.java:3711)
                      at android.app.ActivityThread.performPauseActivity(ActivityThread.java:3688) 
                      at android.app.ActivityThread.performPauseActivity(ActivityThread.java:3662) 
                      at android.app.ActivityThread.handlePauseActivity(ActivityThread.java:3636) 
                      at android.app.ActivityThread.-wrap16(ActivityThread.java) 
                      at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1481) 
                      at android.os.Handler.dispatchMessage(Handler.java:102) 
                      at android.os.Looper.loop(Looper.java:154) 
                      at android.app.ActivityThread.main(ActivityThread.java:6077) 
                      at java.lang.reflect.Method.invoke(Native Method) 
                      at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:865) 
                      at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:755) 
```

See on emulators for API 19 and 24, master and develop branches. I haven't checked others.

--- a/blocklylib-core/src/main/java/com/google/blockly/android/codegen/CodeGeneratorManager.java
+++ b/blocklylib-core/src/main/java/com/google/blockly/android/codegen/CodeGeneratorManager.java
@@ -51,7 +51,7 @@ public class CodeGeneratorManager {
             public void onServiceConnected(ComponentName className, IBinder binder) {
                 try {
                     if (!mResumed) {
-                        mContext.unbindService(mCodeGenerationConnection);
+                        unbind();
                     } else {
                         mGeneratorService = ((CodeGeneratorService.CodeGeneratorBinder) binder).getService();
 
@@ -76,9 +76,7 @@ public class CodeGeneratorManager {
      */
     public void onPause() {
         mResumed = false;
-        if (isBound()) {
-            mContext.unbindService(mCodeGenerationConnection);
-        }
+        unbind();
     }
 
     /**
@@ -88,6 +86,16 @@ public class CodeGeneratorManager {
     public void onResume() {
         mResumed = true;
         mStoredRequests.clear();
+    }
+
+    /**
+     * Checks if the service is currently bound and unbinds it if it is.
+     */
+    protected void unbind() {
+        if (isBound()) {
+            mContext.unbindService(mCodeGenerationConnection);
+            mGeneratorService = null;
+        }
     }
 
     /**

